File: apps/react-node/server/request-snapshot.ts
import type { Request } from "express";

const HEADER_ALLOW = new Set([
  "x-request-id",
  "x-request-origin",
  "x-dashboard-page",
  "x-session-id",
  "content-type",
  "accept",
  "user-agent",
  "host",
]);

export function requestHeaders(req: Request): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [key, value] of Object.entries(req.headers)) {
    const lowered = key.toLowerCase();
    if (!HEADER_ALLOW.has(lowered)) {
      continue;
    }
    if (typeof value === "string") {
      out[lowered] = value;
    } else if (Array.isArray(value) && value.length > 0) {
      out[lowered] = value[0];
    }
  }
  return out;
}

export function requestUrlParams(req: Request): Record<string, unknown> | undefined {
  const query = req.query as Record<string, unknown>;
  return Object.keys(query).length > 0 ? query : undefined;
}

export function requestBody(req: Request): Record<string, unknown> | undefined {
  if (req.method === "GET" || req.method === "HEAD") {
    return undefined;
  }
  if (req.body && typeof req.body === "object" && !Array.isArray(req.body)) {
    const body = req.body as Record<string, unknown>;
    return Object.keys(body).length > 0 ? body : undefined;
  }
  return undefined;
}

File: apps/python/src/exercises/web/request_snapshot.py
"""Safe HTTP request headers and body maps for structured access logs."""

from __future__ import annotations

from typing import Any

from flask import Request

_HEADER_ALLOW = frozenset(
    {
        "x-request-id",
        "x-request-origin",
        "x-dashboard-page",
        "x-session-id",
        "content-type",
        "accept",
        "user-agent",
        "host",
    }
)


def request_headers(req: Request) -> dict[str, str]:
    out: dict[str, str] = {}
    for key, value in req.headers.items():
        lowered = key.lower()
        if lowered in _HEADER_ALLOW:
            out[lowered] = value
    return out


def request_url_params(req: Request) -> dict[str, Any]:
    args = req.args.to_dict(flat=False)
    if not args:
        return {}
    return {
        key: values[0] if len(values) == 1 else values
        for key, values in args.items()
    }


def request_body(req: Request) -> dict[str, Any]:
    if req.method in {"GET", "HEAD", "DELETE", "OPTIONS"}:
        return {}

    data = req.get_json(silent=True)
    if isinstance(data, dict):
        return data
    if isinstance(data, list):
        return {"_json": data}

    form = req.form.to_dict(flat=False)
    if form:
        return {
            key: values[0] if len(values) == 1 else values
            for key, values in form.items()
        }

    return {}
